readme.md
# ApiResponse

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Total Downloads][ico-downloads]][link-downloads]
[![StyleCI][ico-styleci]][link-styleci]

[comment]: <> ([![Build Status][ico-travis]][link-travis])


Class to generate a standard structure for api json responses.

## Installation

Via Composer

``` bash
$ composer require willywes/apiresponse
```

## Usage

#### Imports

``` php
use Willywes\ApiResponse\ApiResponse;
```

### Functions of Control (HTTP/200 OK)

Default functions that always return a http 200 code, but have a control state.

#### Params

|Param          |Description                            |
|---------------|---------------------------------------|
|**data**      |array of data for response (allow null)|
|**message**   |custom message to response (optional)  |
|**title**     |custom title to response (optional)    |

#### Functions

|Function                        |Description                                      |
|--------------------------------|-------------------------------------------------|
|**JsonSuccess**               | Response status "success" with HTTP 200          |
|**JsonError**                 | Response status "error" with HTTP 200            |
|**JsonWarning**               | Response status "warning" with HTTP 200          |
|**JsonInfo**                  | Response status "info" with HTTP 200             |
|**JsonFieldValidation**      | Response status "field_validation" with HTTP 200|

#### Examples

##### Success example

``` php
//Execution in php
return ApiResponse::JsonSuccess([
        'user' => User::first(),
        'roles' => Role::all(),
    ]);
```

``` json
//Response
{
    "status":"success",
    "title":"Operación Exitosa.",
    "message": null,
    "data":{
        "user":{
            "id":1,
            "full_name":"John Smith",
            "email":"jsmith@test.cl",
            "role_id":1
        },
        "roles":[
            {
                "id":1,
                "name":"God Admin"
            },
            {
                "id":2,
                "name":"Administrator"
            }
        ]
    }
}
```

``` json5
//HTTP Response 
Status Code: 200 OK
```

##### Error example

``` php
//Execution in php
return ApiResponse::JsonError(null, 'something has gone wrong!', 'oops');
```

``` json
//Response
{
    "status":"error",
    "title":"oops",
    "message":"something has gone wrong!",
    "data": null
}
```

``` json5
//HTTP Response 
Status Code: 200 OK
```

### Functions with specific HTTP Code

Default functions that returns a specific http code, but in the same way the body responds

#### Params

|Param          |Description                            |
|---------------|---------------------------------------|
|**data**|array of data for response (allow null)|
|**message**|custom message to response (optional)|

#### Functions Code HTTP 1XX

|Http Code|Function Name|Description|
|---|---|---|
|**100**|**Http100 or Continue**|Response status "info" with HTTP 100|
|**101**|**Http101 or SwitchingProtocols**|Response status "info" with HTTP 101|
|**102**|**Http102 or Processing**|Response status "info" with HTTP 102|
|**103**|**Http103 or EarlyHints**|Response status "info" with HTTP 103|

#### Functions Code HTTP 2XX

|Http Code|Function Name|Description|
|---|---|---|
|**200**|**Http200 or Ok**|Response status "success" with HTTP 200|
|**201**|**Http201 or Created**|Response status "success" with HTTP 201|
|**202**|**Http202 or Accepted**|Response status "success" with HTTP 202|
|**203**|**Http203 or NonAuthoritativeInformation**|Response status "success" with HTTP 203|
|**204**|**Http204 or NoContent**|Response status "success" with HTTP 204|
|**205**|**Http205 or ResetContent**|Response status "success" with HTTP 205|
|**206**|**Http206 or PartialContent**|Response status "success" with HTTP 206|

#### Functions Code HTTP 3XX

|Http Code|Function Name|Description|
|---|---|---|
|**300**|**Http300 or MultipleChoices**|Response status "redirection" with HTTP 300|
|**301**|**Http301 or MovedPermanently**|Response status "redirection" with HTTP 301|
|**302**|**Http302 or MovedTemporarily**|Response status "redirection" with HTTP 302|
|**303**|**Http303 or SeeOther**|Response status "redirection" with HTTP 303|
|**304**|**Http304 or NotModified**|Response status "redirection" with HTTP 304|
|**305**|**Http305 or UseProxy**|Response status "redirection" with HTTP 305|
|**307**|**Http307 or TemporaryRedirect**|Response status "redirection" with HTTP 307|
|**308**|**Http308 or PermanentRedirect**|Response status "redirection" with HTTP 308|

#### Functions Code HTTP 4XX

|Http Code|Function Name|Description|
|---|---|---|
|**400**|**Http400 or BadRequest**|Response status "error" with HTTP 400|
|**401**|**Http401 or Unauthorized**|Response status "error" with HTTP 401|
|**402**|**Http402 or PaymentRequired**|Response status "error" with HTTP 402|
|**403**|**Http403 or Forbidden**|Response status "error" with HTTP 403|
|**404**|**Http404 or NotFound**|Response status "error" with HTTP 404|
|**405**|**Http405 or MethodNotAllowed**|Response status "error" with HTTP 405|
|**406**|**Http406 or NotAcceptable**|Response status "error" with HTTP 406|
|**407**|**Http407 or ProxyAuthenticationRequired**|Response status "error" with HTTP 407|
|**408**|**Http408 or RequestTimeout**|Response status "error" with HTTP 408|
|**409**|**Http409 or Conflict**|Response status "error" with HTTP 409|
|**410**|**Http410 or Gone**|Response status "error" with HTTP 410|
|**411**|**Http411 or LengthRequired**|Response status "error" with HTTP 411|
|**412**|**Http412 or PreconditionFailed**|Response status "error" with HTTP 412|
|**413**|**Http413 or RequestEntityTooLarge**|Response status "error" with HTTP 413|
|**414**|**Http414 or RequestUriTooLarge**|Response status "error" with HTTP 414|
|**415**|**Http415 or UnsupportedMediaType**|Response status "error" with HTTP 415|
|**416**|**Http416 or RequestedRangeNotSatisfiable**|Response status "error" with HTTP 416|
|**417**|**Http417 or ExpectationFailed**|Response status "error" with HTTP 417|
|**421**|**Http421 or MissDirectedRequest**|Response status "error" with HTTP 421|
|**422**|**Http422 or UnprocessableEntity**|Response status "error" with HTTP 422|
|**423**|**Http423 or Locked**|Response status "error" with HTTP 423|
|**424**|**Http424 or FailedDependency**|Response status "error" with HTTP 424|
|**426**|**Http426 or UpgradeRequired**|Response status "error" with HTTP 426|
|**428**|**Http428 or PreconditionRequired**|Response status "error" with HTTP 428|
|**429**|**Http429 or TooManyRequests**|Response status "error" with HTTP 429|
|**431**|**Http431 or RequestHeaderFieldsTooLarge**|Response status "error" with HTTP 431|
|**451**|**Http451 or UnavailableForLegalReasons**|Response status "error" with HTTP 451|

#### Functions Code HTTP 5XX

|Http Code|Function Name|Description|
|---|---|---|
|**500**|**Http500 or InternalServerError**|Response status "error" with HTTP 500|
|**501**|**Http501 or NotImplemented**|Response status "error" with HTTP 501|
|**502**|**Http502 or BadGateway**|Response status "error" with HTTP 502|
|**503**|**Http503 or ServiceUnavailable**|Response status "error" with HTTP 503|
|**504**|**Http504 or GatewayTimeout**|Response status "error" with HTTP 504|
|**505**|**Http505 or HTTPVersionNotSupported**|Response status "error" with HTTP 505|
|**506**|**Http506 or VariantAlsoNegotiates**|Response status "error" with HTTP 506|
|**507**|**Http507 or InsufficientStorage**|Response status "error" with HTTP 507|
|**508**|**Http508 or LoopDetected**|Response status "error" with HTTP 508|
|**510**|**Http510 or NotExtended**|Response status "error" with HTTP 510|
|**511**|**Http511 or NetworkAuthenticationRequired**|Response status "error" with HTTP 511|

#### Examples

##### 404 Error example

``` php
//Execution in php
return ApiResponse::NotFound(null, 'object not found!');
// or
//Execution in php without params
return ApiResponse::NotFound();
// or
return ApiResponse::Http404();
```

``` json
//Response
{
   "status": "error",
   "message": "Not Found",
   "data": null
}
```

``` json5
//HTTP Response 
Status Code: 404 Not Found
```

##### 401 Error example

``` php
//Execution in php
return ApiResponse::Unauthorized();
// or
return ApiResponse::Http401();
```

``` json
//Response
{
   "status": "error",
   "message": "Unauthorized",
   "data": null
}
```

``` json5
//HTTP Response 
Status Code: 401 Unauthorized
```

##### 403 Error example

``` php
//Execution in php
return ApiResponse::Forbidden();
// or
return ApiResponse::Http403();
```

``` json
//Response
{
   "status": "error",
   "message": "Forbidden",
   "data": null
}
```

``` json5
//HTTP Response 
Status Code: 403 Forbidden
```

## Credits

- [Alejandro Isla][link-author]

## License

license. Please see the [license file](license.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/willywes/apiresponse.svg?style=flat-square

[ico-downloads]: https://img.shields.io/packagist/dt/willywes/apiresponse.svg?style=flat-square

[ico-travis]: https://img.shields.io/travis/willywes/apiresponse/master.svg?style=flat-square

[ico-styleci]: https://styleci.io/repos/306464936/shield

[link-packagist]: https://packagist.org/packages/willywes/apiresponse

[link-downloads]: https://packagist.org/packages/willywes/apiresponse

[link-travis]: https://travis-ci.org/willywes/apiresponse

[link-styleci]: https://styleci.io/repos/306464936

[link-author]: https://github.com/willywes


changelog.md
# Changelog

All notable changes to `ApiResponse` will be documented in this file.
## v1.7.0 - 2023-09-11
### Added
- Illuminate Support ~10

## v1.6.1 - 2022-04-25
### Added
- Illuminate Support ~9

## v1.6.0 - 2022-01-21
### Added
- JsonFieldValidationError with http code 422

## v1.5.0 - 2021-11-24
### Fix
- All status code response

## v1.4.0 - 2021-11-08
### Added
- All responses http code functions

## v1.3.0 - 2021-08-06
### Added
- All responses code
---
## v1.2.0 - 2021-08-06
### Fixed
- HTTP 2.x.x response success
---
## v1.0.0 - 2021-07-20
### Added
- New HTTP Codes
---
## Version 0.2.5
### Added
- Everything
---
## Version 0.2.1
### Added
- Everything


src/ICode.php
<?php

namespace Willywes\ApiResponse;

interface ICode
{
    // HTML CODES
    const CODE_100 = 'Continue';
    const CODE_101 = 'Switching Protocols';
    const CODE_102 = 'Processing';
    const CODE_103 = 'Early Hints';
    const CODE_200 = 'OK';
    const CODE_201 = 'Created';
    const CODE_202 = 'Accepted';
    const CODE_203 = 'Non-Authoritative Information';
    const CODE_204 = 'No Content';
    const CODE_205 = 'Reset Content';
    const CODE_206 = 'Partial Content';
    const CODE_300 = 'Multiple Choices';
    const CODE_301 = 'Moved Permanently';
    const CODE_302 = 'Moved Temporarily';
    const CODE_303 = 'See Other';
    const CODE_304 = 'Not Modified';
    const CODE_305 = 'Use Proxy';
    const CODE_306 = 'UnUsed'; //DEPRECATED
    const CODE_307 = 'TemporaryRedirect';
    const CODE_308 = 'PermanentRedirect';
    const CODE_400 = 'Bad Request';
    const CODE_401 = 'Unauthorized';
    const CODE_402 = 'Payment Required';
    const CODE_403 = 'Forbidden';
    const CODE_404 = 'Not Found';
    const CODE_405 = 'Method Not Allowed';
    const CODE_406 = 'Not Acceptable';
    const CODE_407 = 'Proxy Authentication Required';
    const CODE_408 = 'Request Time-out';
    const CODE_409 = 'Conflict';
    const CODE_410 = 'Gone';
    const CODE_411 = 'Length Required';
    const CODE_412 = 'Precondition Failed';
    const CODE_413 = 'Request Entity Too Large';
    const CODE_414 = 'Request-URI Too Large';
    const CODE_415 = 'Unsupported Media Type';
    const CODE_416 = 'Requested Range Not Satisfiable';
    const CODE_417 = 'Expectation Failed';
    const CODE_421 = 'Miss Directed Request';
    const CODE_422 = 'Unprocessable Entity';
    const CODE_423 = 'Locked';
    const CODE_424 = 'Failed Dependency';
    const CODE_426 = 'Upgrade Required';
    const CODE_428 = 'Precondition Required';
    const CODE_429 = 'Too Many Requests';
    const CODE_431 = 'Request Header Fields Too Large';
    const CODE_451 = 'Unavailable For Legal Reasons';
    const CODE_500 = 'Internal Server Error';
    const CODE_501 = 'Not Implemented';
    const CODE_502 = 'Bad Gateway';
    const CODE_503 = 'Service Unavailable';
    const CODE_504 = 'Gateway Time-out';
    const CODE_505 = 'HTTP Version not supported';
    const CODE_506 = 'Variant Also Negotiates';
    const CODE_507 = 'Insufficient Storage';
    const CODE_508 = 'Loop Detected';
    const CODE_510 = 'Not Extended';
    const CODE_511 = 'Network Authentication Required';

    // DEFAULT STATUS

    const STATUS_INFO = 'info';
    const STATUS_SUCCESS = 'success';
    const STATUS_WARNING = 'warning';
    const STATUS_ERROR = 'error';
    const STATUS_FIELDS_VALIDATION = 'fields_validation';
    const STATUS_REDIRECTION = 'redirection';


}


src/Format.php
<?php

namespace Willywes\ApiResponse;

class Format implements ICode
{
    public static function Response($status, $title = null, $message = null, $data = null): array
    {
        return [
            'status' => $status,
            'title' => $title ?? self::SetDefaultTitle($status),
            'message' => $message,
            'data' => $data,
        ];
    }

    private static function SetDefaultTitle($class): string
    {
        switch ($class) {
            case self::STATUS_INFO:
                return 'Info';
            case self::STATUS_SUCCESS:
                return 'Success';
            case self::STATUS_WARNING:
                return 'Warning';
            case self::STATUS_ERROR:
                return 'Error';
            case self::STATUS_REDIRECTION:
                return 'Redirection';
            case self::STATUS_FIELDS_VALIDATION:
                return 'Error Fields';
            default:
                return '';
        }

    }
}


src/ApiResponse.php
<?php

namespace Willywes\ApiResponse;

class ApiResponse implements ICode
{
    public static function JsonInfo($data = null, $message = null, $title = null)
    {
        return self::responseJson(Format::Response(self::STATUS_INFO, $title, $message, $data), 200);
    }

    public static function JsonWarning($data = null, $message = null, $title = null)
    {
        return self::responseJson(Format::Response(self::STATUS_WARNING, $title, $message, $data), 200);
    }

    public static function JsonSuccess($data = null, $message = null, $title = null)
    {
        return self::responseJson(Format::Response(self::STATUS_SUCCESS, $title, $message, $data), 200);
    }

    public static function JsonError($data = null, $message = null, $title = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, $title, $message, $data), 200);
    }

    public static function JsonFieldValidation($data = null, $message = null, $title = null)
    {
        return self::responseJson(Format::Response(self::STATUS_FIELDS_VALIDATION, $title, $message, $data), 200);
    }

    // HTTPS RESPONSE

    //INFO 10*
    public static function Http100($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_INFO, self::CODE_100, $message, $data), 100);
    }

    public static function Http101($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_INFO, self::CODE_101, $message, $data), 101);
    }

    public static function Http102($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_INFO, self::CODE_102, $message, $data), 102);
    }

    public static function Http103($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_INFO, self::CODE_103, $message, $data), 103);
    }

    public static function Continue($data = null, $message = null)
    {
        return self::Http100($data, $message);
    }

    public static function SwitchingProtocols($data = null, $message = null)
    {
        return self::Http101($data, $message);
    }

    public static function Processing($data = null, $message = null)
    {
        return self::Http102($data, $message);
    }

    public static function EarlyHints($data = null, $message = null)
    {
        return self::Http103($data, $message);
    }

    // ACCEPTS 20*
    public static function Http200($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_SUCCESS, self::CODE_200, $message, $data), 200);
    }

    public static function Http201($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_SUCCESS, self::CODE_201, $message, $data), 201);
    }

    public static function Http202($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_SUCCESS, self::CODE_202, $message, $data), 202);
    }

    public static function Http203($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_SUCCESS, self::CODE_203, $message, $data), 203);
    }

    public static function Http204($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_SUCCESS, self::CODE_204, $message, $data), 204);
    }

    public static function Http205($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_SUCCESS, self::CODE_205, $message, $data), 205);
    }

    public static function Http206($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_SUCCESS, self::CODE_206, $message, $data), 206);
    }

    public static function Ok($data = null, $message = null)
    {
        return self::Http200($data, $message);
    }

    public static function Created($data = null, $message = null)
    {
        return self::Http201($data, $message);
    }

    public static function Accepted($data = null, $message = null)
    {
        return self::Http202($data, $message);
    }

    public static function NonAuthoritativeInformation($data = null, $message = null)
    {
        return self::Http203($data, $message);
    }

    public static function NoContent($data = null, $message = null)
    {
        return self::Http204($data, $message);
    }

    public static function ResetContent($data = null, $message = null)
    {
        return self::Http205($data, $message);
    }

    public static function PartialContent($data = null, $message = null)
    {
        return self::Http206($data, $message);
    }

    //REDIRECTION 30*
    public static function Http300($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_REDIRECTION, self::CODE_300, $message, $data), 300);
    }

    public static function Http301($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_REDIRECTION, self::CODE_301, $message, $data), 301);
    }

    public static function Http302($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_REDIRECTION, self::CODE_302, $message, $data), 302);
    }

    public static function Http303($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_REDIRECTION, self::CODE_303, $message, $data), 303);
    }

    public static function Http304($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_REDIRECTION, self::CODE_304, $message, $data), 304);
    }

    public static function Http305($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_REDIRECTION, self::CODE_305, $message, $data), 305);
    }

    public static function Http307($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_REDIRECTION, self::CODE_307, $message, $data), 307);
    }

    public static function Http308($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_REDIRECTION, self::CODE_308, $message, $data), 308);
    }

    public static function MultipleChoices($data = null, $message = null)
    {
        return self::Http300($data, $message);
    }

    public static function MovedPermanently($data = null, $message = null)
    {
        return self::Http301($data, $message);
    }

    public static function MovedTemporarily($data = null, $message = null)
    {
        return self::Http302($data, $message);
    }

    public static function SeeOther($data = null, $message = null)
    {
        return self::Http303($data, $message);
    }

    public static function NotModified($data = null, $message = null)
    {
        return self::Http304($data, $message);
    }

    public static function UseProxy($data = null, $message = null)
    {
        return self::Http305($data, $message);
    }

    public static function TemporaryRedirect($data = null, $message = null)
    {
        return self::Http307($data, $message);
    }

    public static function PermanentRedirect($data = null, $message = null)
    {
        return self::Http308($data, $message);
    }

    // ERRORS 40*

    public static function Http400($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_400, $message, $data), 400);
    }

    public static function Http401($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_401, $message, $data), 401);
    }

    public static function Http402($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_402, $message, $data), 402);
    }

    public static function Http403($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_403, $message, $data), 403);
    }

    public static function Http404($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_404, $message, $data), 404);
    }

    public static function Http405($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_405, $message, $data), 405);
    }

    public static function Http406($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_406, $message, $data), 406);
    }

    public static function Http407($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_407, $message, $data), 407);
    }

    public static function Http408($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_408, $message, $data), 408);
    }

    public static function Http409($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_409, $message, $data), 409);
    }

    public static function Http410($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_410, $message, $data), 410);
    }

    public static function Http411($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_411, $message, $data), 411);
    }

    public static function Http412($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_412, $message, $data), 412);
    }

    public static function Http413($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_413, $message, $data), 413);
    }

    public static function Http414($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_414, $message, $data), 414);
    }

    public static function Http415($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_415, $message, $data), 415);
    }

    public static function Http416($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_416, $message, $data), 416);
    }

    public static function Http417($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_417, $message, $data), 417);
    }

    public static function Http421($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_421, $message, $data), 421);
    }

    public static function Http422($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_422, $message, $data), 422);
    }

    public static function Http423($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_423, $message, $data), 423);
    }

    public static function Http424($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_424, $message, $data), 424);
    }

    public static function Http426($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_426, $message, $data), 426);
    }

    public static function Http428($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_428, $message, $data), 428);
    }

    public static function Http429($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_429, $message, $data), 429);
    }

    public static function Http431($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_431, $message, $data), 431);
    }

    public static function Http451($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_451, $message, $data), 451);
    }

    public static function BadRequest($data = null, $message = null)
    {
        return self::Http400($data, $message);
    }

    public static function Unauthorized($data = null, $message = null)
    {
        return self::Http401($data, $message);
    }

    public static function PaymentRequired($data = null, $message = null)
    {
        return self::Http402($data, $message);
    }

    public static function Forbidden($data = null, $message = null)
    {
        return self::Http403($data, $message);
    }

    public static function NotFound($data = null, $message = null)
    {
        return self::Http404($data, $message);
    }

    public static function MethodNotAllowed($data = null, $message = null)
    {
        return self::Http405($data, $message);
    }

    public static function NotAcceptable($data = null, $message = null)
    {
        return self::Http406($data, $message);
    }

    public static function ProxyAuthenticationRequired($data = null, $message = null)
    {
        return self::Http407($data, $message);
    }

    public static function RequestTimeout($data = null, $message = null)
    {
        return self::Http408($data, $message);
    }

    public static function Conflict($data = null, $message = null)
    {
        return self::Http409($data, $message);
    }

    public static function Gone($data = null, $message = null)
    {
        return self::Http410($data, $message);
    }

    public static function LengthRequired($data = null, $message = null)
    {
        return self::Http411($data, $message);
    }

    public static function PreconditionFailed($data = null, $message = null)
    {
        return self::Http412($data, $message);
    }

    public static function RequestEntityTooLarge($data = null, $message = null)
    {
        return self::Http413($data, $message);
    }

    public static function RequestUriTooLarge($data = null, $message = null)
    {
        return self::Http414($data, $message);
    }

    public static function UnsupportedMediaType($data = null, $message = null)
    {
        return self::Http415($data, $message);
    }

    public static function RequestedRangeNotSatisfiable($data = null, $message = null)
    {
        return self::Http416($data, $message);
    }

    public static function ExpectationFailed($data = null, $message = null)
    {
        return self::Http417($data, $message);
    }

    public static function MissDirectedRequest($data = null, $message = null)
    {
        return self::Http421($data, $message);
    }

    public static function UnprocessableEntity($data = null, $message = null)
    {
        return self::Http422($data, $message);
    }

    public static function Locked($data = null, $message = null)
    {
        return self::Http423($data, $message);
    }

    public static function FailedDependency($data = null, $message = null)
    {
        return self::Http424($data, $message);
    }

    public static function UpgradeRequired($data = null, $message = null)
    {
        return self::Http426($data, $message);
    }

    public static function PreconditionRequired($data = null, $message = null)
    {
        return self::Http428($data, $message);
    }

    public static function TooManyRequests($data = null, $message = null)
    {
        return self::Http429($data, $message);
    }

    public static function RequestHeaderFieldsTooLarge($data = null, $message = null)
    {
        return self::Http431($data, $message);
    }

    public static function UnavailableForLegalReasons($data = null, $message = null)
    {
        return self::Http451($data, $message);
    }

    // ERRORS SERVER 50*

    public static function Http500($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_500, $message, $data), 500);
    }

    public static function Http501($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_501, $message, $data), 501);
    }

    public static function Http502($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_502, $message, $data), 502);
    }

    public static function Http503($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_503, $message, $data), 503);
    }

    public static function Http504($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_504, $message, $data), 504);
    }

    public static function Http505($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_505, $message, $data), 505);
    }

    public static function Http506($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_506, $message, $data), 506);
    }

    public static function Http507($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_507, $message, $data), 507);
    }

    public static function Http508($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_508, $message, $data), 508);
    }

    public static function Http510($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_510, $message, $data), 510);
    }

    public static function Http511($data = null, $message = null)
    {
        return self::responseJson(Format::Response(self::STATUS_ERROR, self::CODE_511, $message, $data), 511);
    }

    public static function InternalServerError($data = null, $message = null)
    {
        return self::Http500($data, $message);
    }

    public static function NotImplemented($data = null, $message = null)
    {
        return self::Http501($data, $message);
    }

    public static function BadGateway($data = null, $message = null)
    {
        return self::Http502($data, $message);
    }

    public static function ServiceUnavailable($data = null, $message = null)
    {
        return self::Http503($data, $message);
    }

    public static function GatewayTimeout($data = null, $message = null)
    {
        return self::Http504($data, $message);
    }

    public static function HTTPVersionNotSupported($data = null, $message = null)
    {
        return self::Http505($data, $message);
    }

    public static function VariantAlsoNegotiates($data = null, $message = null)
    {
        return self::Http506($data, $message);
    }

    public static function InsufficientStorage($data = null, $message = null)
    {
        return self::Http507($data, $message);
    }

    public static function LoopDetected($data = null, $message = null)
    {
        return self::Http508($data, $message);
    }

    public static function NotExtended($data = null, $message = null)
    {
        return self::Http510($data, $message);
    }

    public static function NetworkAuthenticationRequired($data = null, $message = null)
    {
        return self::Http511($data, $message);
    }

    private static function responseJson($response, $httpCode)
    {
        return response()->json($response, $httpCode);
    }
}
